### PR TITLE
ParameterizeChartsInHeap Parallelized

### PR DESCRIPTION
**Pull request by WurmD**
James30K                                     no_parallel		parallel									
			QUALITY				:         66s	                  55s	
			FAST				:	   5.5s			   3.5s
		John400K
			QUALITY				:	3650s			2277s
			FAST				:	 142s			 124s

Some improvements with this version. All tests in test suite pass

### CODE REVIEW

File: UVAtlas/isochart/isochartengine.cpp [hunk 1]
```diff
@@ -274,6 +278,125 @@ HRESULT CIsochartEngine::InitializeCurrentChartHeap()
...
+    /// 3rd children = parent, goto 2nd 
+    std::vector<CIsochartMesh*> parent;
+    while (!m_currentChartHeap.empty())
+        parent.emplace_back(m_currentChartHeap.cutTopData());
+
+    HRESULT hrOut = S_OK;
+    int runs = 0;
+    while (!parent.empty() && !FAILED(hrOut))
+    {
+        // std::cout << "Run " << runs++ << " parent size: " << parent.size() << " :";
```
Comment (walbourn) [inline UVAtlas/isochart/isochartengine.cpp]: Delete test comments

File: UVAtlas/isochart/isochartengine.cpp [hunk 2]
```diff
@@ -274,6 +278,125 @@ HRESULT CIsochartEngine::InitializeCurrentChartHeap()
...
+#pragma omp critical
+                        m_finalChartList.push_back(pChart);
+                    }
+                    catch (std::bad_alloc&)
+                    {
+                        hrOut = E_OUTOFMEMORY;
+                    }
+                }
+
+                std::chrono::duration<double> elapsed_seconds2 = std::chrono::system_clock::now() - start;
```
Comment (walbourn) [inline UVAtlas/isochart/isochartengine.cpp]: Delete test code.

File: UVAtlas/isochart/isochartengine.cpp [hunk 3]
```diff
@@ -545,7 +668,9 @@ HRESULT CIsochartEngine::PartitionByGlobalAvgL2Stretch(
     {
         // 3.1. Generate initial parameterization for charts in current chart heap
         FAILURE_RETURN(
-            ParameterizeChartsInHeap(bCountParition, MaxChartNumber));
+            ParameterizeChartsInHeap(bCountParition, MaxChartNumber)
+            // ParameterizeChartsInHeapParallelized(bCountParition, MaxChartNumber)
```
Comment (walbourn) [inline UVAtlas/isochart/isochartengine.cpp]: You aren't using it here... Is that by design?

Comment (walbourn) [reply UVAtlas/isochart/isochartengine.cpp]: I'd suggest doing:

```
#ifdef _OPENMP
ParameterizeChartsInHeapParallelized(bCountParition, MaxChartNumber)
#else
ParameterizeChartsInHeap(bCountParition, MaxChartNumber)
#endif
```

```

File: UVAtlas/isochart/isochartengine.cpp [hunk 4]
```diff
@@ -274,6 +278,125 @@ HRESULT CIsochartEngine::InitializeCurrentChartHeap()
     return S_OK;
 }
 
+
+HRESULT CIsochartEngine::ParameterizeChartsInHeapParallelized(
```
Comment (walbourn) [inline UVAtlas/isochart/isochartengine.cpp]: Use ``#ifdef _OPENMP`` around this method in the cpp and the header.

File: UVAtlas/isochart/isochartengine.cpp [hunk 5]
```diff
@@ -274,6 +278,125 @@ HRESULT CIsochartEngine::InitializeCurrentChartHeap()
...
+    while (!parent.empty() && !FAILED(hrOut))
+    {
+        // std::cout << "Run " << runs++ << " parent size: " << parent.size() << " :"; 
+        std::vector<CIsochartMesh*> children;
+#pragma omp parallel 
+        {
+
+            std::vector<CIsochartMesh*> children_thrd;
+#pragma omp for
+            for (int n = 0; n < parent.size(); ++n)
```
Comment (walbourn) [inline UVAtlas/isochart/isochartengine.cpp]: ``static_cast<int>(parent.size());``

File: UVAtlas/isochart/isochartengine.cpp [hunk 6]
```diff
@@ -11,6 +11,10 @@
 #include "isochartengine.h"
 #include "isochart.h"
 #include "isochartmesh.h"
+#include <chrono>
```
Comment (walbourn) [inline UVAtlas/isochart/isochartengine.cpp]: Don't need these once you delete the test/timing code

File: UVAtlas/isochart/isochartengine.h [hunk 7]
```diff
@@ -122,6 +122,7 @@ class CIsochartEngine : public IIsochartEngine
         float& MaxChartStretchOut,
         uint32_t* pFaceAttributeIDOut);
 
+    HRESULT ParameterizeChartsInHeapParallelized(bool bFirstTime, size_t MaxChartNumber);
```
Comment (walbourn) [inline UVAtlas/isochart/isochartengine.h]: Again ``#ifdef _OPENMP`` around this method's prototype.

File: UVAtlas/isochart/isochartengine.cpp (hunk 5 above)
Comment (WurmD) [reply UVAtlas/isochart/isochartengine.cpp]: done

File: UVAtlas/isochart/isochartengine.h (hunk 7 above)
Comment (WurmD) [reply UVAtlas/isochart/isochartengine.h]: done

**Comment by WurmD** [comment]
There. All suggestions implemented. Tests still run and pass.
Also some style fixes as well.

**Pull request merged**